### ISSUE TITLE
Commit-based Git downloads fail checksum when cached

### ISSUE THREAD

**Issue by HadrienG2** [spack/spack#8756]
When using commit-based Git downloads (those of the form `version('x.y.z', commit='abcdef0123456789')`), the checksum test fails if Spack has already cached this specific commit from the repository.

### Expected Result

If Spack was happy initially downloading a Git commit, it should remain happy when reusing a cached version of this commit.

### Actual Result

First build of a commit-based package:

    $ spack install acts-core@0.6.0
    ==> bzip2 is already installed in /home/hadrien/Software/spack/opt/spack/linux-opensuse_tumbleweed20180712-x86_64/gcc-8.1.1/bzip2-1.0.6-62ez6e6lkdgaqxpfyen44zqzgg5j5qy2
       [...]
    ==> Installing acts-core
    ==> Cloning git repository: https://gitlab.cern.ch/acts/acts-core.git at commit 7358bc8b274c5c474c98c9e44dd796666de11a9d
    ==> No checksum needed when fetching with git
    ==> Already staged acts-core-0.6.0-e2w34654hwu5jfylkyfaxcvxvcl5577j in /home/hadrien/Software/spack/var/spack/stage/acts-core-0.6.0-e2w34654hwu5jfylkyfaxcvxvcl5577j
    ==> No patches needed for acts-core
        [...]

Second build of a commit-based package:

    $ spack uninstall -y acts-core@0.6.0 && spack install acts-core@0.6.0
    ==> Successfully uninstalled acts-core@0.6.0%gcc@8.1.1 build_type=RelWithDebInfo ~examples~integration_tests+legacy~material_plugin~tests arch=linux-opensuse_tumbleweed20180712-x86_64 /e2w3465
    ==> bzip2 is already installed in /home/hadrien/Software/spack/opt/spack/linux-opensuse_tumbleweed20180712-x86_64/gcc-8.1.1/bzip2-1.0.6-62ez6e6lkdgaqxpfyen44zqzgg5j5qy2
       [...]
    ==> Installing acts-core
    ==> Using cached archive: /home/hadrien/Software/spack/var/spack/cache/acts-core/acts-core-0.6.0.tar.gz
    ==> Warning: Fetching from mirror without a checksum!
      This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
    ==> Staging archive: /home/hadrien/Software/spack/var/spack/stage/acts-core-0.6.0-e2w34654hwu5jfylkyfaxcvxvcl5577j/acts-core-0.6.0.tar.gz
    ==> Created stage in /home/hadrien/Software/spack/var/spack/stage/acts-core-0.6.0-e2w34654hwu5jfylkyfaxcvxvcl5577j
    ==> No patches needed for acts-core
        [...]

Notice that there is a checksum warning on the second run

### Steps to reproduce the issue

1. Put a commit-based version in a package specification, e.g.
    ```python
    version('0.6.0',
            git='https://gitlab.cern.ch/acts/acts-core.git',
            commit='7358bc8b274c5c474c98c9e44dd796666de11a9d')
    ```
2. Install the version once. This should succeed.
3. Uninstall the version and install it to use the cache. This should trigger the warning above.

### Information on your system

As far as I can tell, this behaviour comes from a system-independent part of Spack.

**Comment by adamjstewart** [comment]
Is this a duplicate of #8746?

**Comment by HadrienG2** [comment]
I don't think so. #8746 is about git-originated packages failing to build due to missing a `.git` directory, whereas this is about cached git packages building with warnings because they are checksummed differently than fresh downloads.
